src/auth/jwtclient.ts

```diff
@@ -122,44 +122,63 @@ export class JWT extends OAuth2Client implements IdTokenProvider {
   protected async getRequestMetadataAsync(
     url?: string | null
   ): Promise<RequestMetadataResponse> {
-    if (!this.apiKey && !this.hasUserScopes() && url) {
-      if (
-        this.additionalClaims &&
-        (
-          this.additionalClaims as {
-            target_audience: string;
+    if (this.useJWTAccessAlways) {
```

Comment:
This is interesting logic here because it's saying that for clients which set `useJWTAccessAlways` to `true`, do not allow API Key auth or ID token auth. I'm not sure this is what we want to go so far as to say that. The more accurate way (I think) would be something like this:

```js
var canSelfSign = (this.hasUserScopes() && this.useJWTAccessAlways) // self-signed JWT with scopes
  || (this.hasDefaultScopes() && url) // self-signed JWT with audience
if (!this.apiKey && canSelfSign) {
```

Where it used to have:

```js
if (!this.apiKey && !this.hasUserScopes() && url) {
```

Reply:
@bshaffer, I gave this a try earlier today and tests fail, both the ones we've added and the current functionality.

I think this is because we check later if we should sign with audience or scopes, in `jwtaccess.ts`, line 99:

```
if (scopes) {
      defaultClaims = {
        iss: this.email,
        sub: this.email,
        scope: scopes,
        exp,
        iat,
      };
    } else {
      // If either scopes are not specified, or if audience is something unique,
      // just use default service endpoint.
      defaultClaims = {
        iss: this.email,
        sub: this.email,
        aud: url,
        exp,
        iat,
      };
    }
```

> This is interesting logic here because it's saying that for clients which set useJWTAccessAlways to true, do not allow API Key auth or ID token auth

I'm curious about this, isn't a part of this RFC to always use self-signed JWT if it's coming from a GAPIC client?